backend/src/main/java/org/eventplanner/events/ParticipationNotificationUseCase.java

```diff
@@ -0,0 +1,167 @@
+                    .forEach(registration -> userKeyAccessKeyMap.put(registration.getUserKey(), registration));
+
+            var users = event.getRegistrations().stream()
+                    .filter(registration -> registrationKeys.contains(registration.getKey()))
+                    .map(Registration::getUserKey)
+                    .filter(Objects::nonNull)
+                    .map(userService::getUserByKey)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .toList();
+
+            // make sure every userKey has a registration with accessKey
+            // needed for legacy registrations, where accessKey was not generated
+            userKeyAccessKeyMap.entrySet().stream()
+                    .filter(entry -> entry.getValue().getAccessKey() == null)
+                    .forEach(entry -> entry.getValue().setAccessKey(Registration.generateAccessKey()));
+
+            users.forEach(user -> function.accept(user, event, userKeyAccessKeyMap.get(user.getKey())));
+
+            event.setParticipationConfirmationsRequestsSent(alreadySentRequests + 1);
+            eventRepository.update(event);
+        });
+    }
+
+    private List<Event> getEventsToNotify(int alreadySentRequests) {
+        var currentYear = Instant.now().atZone(ZoneId.of("Europe/Berlin")).getYear();
+        var nextYear = currentYear + 1;
+
+        var getEvents = Stream.concat(
+                eventRepository.findAllByYear(currentYear).stream(),
+                eventRepository.findAllByYear(nextYear).stream()
+        ).toList();
+
+        return switch (alreadySentRequests) {
+            case 0 -> getEvents.stream()
+                    .filter(event -> event.getParticipationConfirmationsRequestsSent() == 0)
+                    // filter for 2 weeks remaining
+                    .filter(event -> event.getStart().minusWeeks(2).isBefore(Instant.now().atZone(ZoneId.of("Europe/Berlin"))))
+                    .toList();
+            case 1 -> getEvents.stream()
+                    .filter(event -> event.getParticipationConfirmationsRequestsSent() == 1)
+                    // filter for 1 week remaining
+                    .filter(event -> event.getStart().minusWeeks(1).isBefore(Instant.now().atZone(ZoneId.of("Europe/Berlin"))))
+                    .toList();
+            default -> List.of();
+        };
+    }
+
+    public void confirmRegistration(EventKey eventKey, RegistrationKey registrationKey, String accessKey) {
+        var event = eventRepository.findByKey(eventKey).orElseThrow();
+        var registration = event.getRegistrations().stream()
+                .filter(r -> registrationKey.equals(r.getKey()) && accessKey.equals(r.getAccessKey()))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Registration not found"));
+
+        if (registration.getConfirmedAt() != null) {
```

Comment:
actually we could also overwrite the confirm date, so we have the most recent confirmation date

Reply:
But why?